gnuradio-runtime/lib/buffer.cc

```diff
@@ -263,92 +216,57 @@ void buffer::prune_tags(uint64_t max_time)
 
-unsigned buffer_reader::sample_delay() const { return d_attr_delay; }
-
-int buffer_reader::items_available() const
-{
-    return d_buffer->index_sub(d_buffer->d_write_index, d_read_index);
+    // Mark the callback flag inactive and notify anyone waiting
+    d_callback_flag = false;
+    d_cv.notify_all();
 }
 
-const void* buffer_reader::read_pointer()
-{
-    return &d_buffer->d_base[d_read_index * d_buffer->d_sizeof_item];
-}
+long buffer_ncurrently_allocated() { return s_buffer_count; }
 
-void buffer_reader::update_read_pointer(int nitems)
+std::ostream& operator<<(std::ostream& os, const buffer& buf)
 {
-    gr::thread::scoped_lock guard(*mutex());
-    d_read_index = d_buffer->index_add(d_read_index, nitems);
-    d_abs_read_offset += nitems;
+    os << std::endl
+       << "  sz:    " << buf.d_bufsize << std::endl
+       << "  nrdrs: " << buf.d_readers.size() << std::endl;
+    for (auto& rdr : buf.d_readers) {
+        os << "    rd_idx: " << rdr->get_read_index() << std::endl
+           << "    abs_rd_offset: " << rdr->get_abs_read_offset() << std::endl
+           << std::endl;
+    }
+    return os;
 }
 
-void buffer_reader::get_tags_in_range(std::vector<tag_t>& v,
-                                      uint64_t abs_start,
-                                      uint64_t abs_end,
-                                      long id)
+void buffer::set_context(const buffer_context& context)
 {
-    gr::thread::scoped_lock guard(*mutex());
-
-    uint64_t lower_bound = abs_start - d_attr_delay;
-    // check for underflow and if so saturate at 0
-    if (lower_bound > abs_start)
-        lower_bound = 0;
-    uint64_t upper_bound = abs_end - d_attr_delay;
-    // check for underflow and if so saturate at 0
-    if (upper_bound > abs_end)
-        upper_bound = 0;
-
-    v.clear();
-    std::multimap<uint64_t, tag_t>::iterator itr =
-        d_buffer->get_tags_lower_bound(lower_bound);
-    std::multimap<uint64_t, tag_t>::iterator itr_end =
-        d_buffer->get_tags_upper_bound(upper_bound);
-
-    uint64_t item_time;
-    while (itr != itr_end) {
-        item_time = (*itr).second.offset + d_attr_delay;
-        if ((item_time >= abs_start) && (item_time < abs_end)) {
-            std::vector<long>::iterator id_itr;
-            id_itr = std::find(
-                itr->second.marked_deleted.begin(), itr->second.marked_deleted.end(), id);
-            // If id is not in the vector of marked blocks
-            if (id_itr == itr->second.marked_deleted.end()) {
-                tag_t t = (*itr).second;
-                t.offset += d_attr_delay;
-                v.push_back(t);
-                v.back().marked_deleted.clear();
-            }
-        }
-        itr++;
+    if ((d_context == buffer_context::DEFAULT_INVALID) || (d_context == context)) {
+        // Set the context if the existing value is the default or if it is the
+        // same as what's already been set
+        d_context = context;
```

Comment:
why would it be set multiple times, and that be ok?

Reply:
Context would be set multiple times if an upstream block was connected to multiple downstream blocks.  For now the logic allows context to be set multiple times as long as the value isn't changed.  For now this prevents the case where an upstream host block is connected to both another host block and an accelerated block.  Using this paradigm it might be possible to handle at least some variations of this scenario however I haven't gotten there yet and want to make sure all the basic cases work smoothly before trying out more advanced cases.